omniscidb/QueryBuilder/QueryBuilder.cpp

```diff
@@ -0,0 +1,3072 @@
+    throw InvalidQueryError() << "Cannot parse date add field: '" << field << "'";
+  }
+  return field_names.at(canonical);
+}
+
+JoinType parseJoinType(const std::string& join_type) {
+  static const std::unordered_map<std::string, JoinType> join_types = {
+      {"left", JoinType::LEFT},
+      {"inner", JoinType::INNER},
+      {"semi", JoinType::SEMI},
+      {"anti", JoinType::ANTI}};
+  auto canonical = boost::trim_copy(boost::to_lower_copy(join_type));
+  if (!join_types.count(canonical)) {
+    throw InvalidQueryError() << "Cannot parse join type: '" << join_type << "'";
+  }
+  return join_types.at(canonical);
+}
+
+}  // namespace
+
+BuilderExpr::BuilderExpr(const QueryBuilder* builder,
+                         ExprPtr expr,
+                         const std::string& name,
+                         bool auto_name)
+    : builder_(builder), expr_(expr), name_(name), auto_name_(auto_name) {}
+
+BuilderExpr BuilderExpr::rename(const std::string& name) const {
+  return {builder_, expr_, name, false};
+}
+
+BuilderExpr BuilderExpr::avg() const {
+  if (!expr_->type()->isNumber()) {
+    throw InvalidQueryError() << "Unsupported type for avg aggregate: "
+                              << expr_->type()->toString();
+  }
+  auto agg =
+      makeExpr<AggExpr>(builder_->ctx_.fp64(), AggType::kAvg, expr_, false, nullptr);
+  auto name = name_.empty() ? "avg" : name_ + "_avg";
+  return {builder_, agg, name, true};
+}
+
+BuilderExpr BuilderExpr::min() const {
+  if (!expr_->type()->isNumber() && !expr_->type()->isDateTime()) {
+    throw InvalidQueryError() << "Unsupported type for min aggregate: "
+                              << expr_->type()->toString();
+  }
+  auto agg = makeExpr<AggExpr>(expr_->type(), AggType::kMin, expr_, false, nullptr);
+  auto name = name_.empty() ? "min" : name_ + "_min";
+  return {builder_, agg, name, true};
+}
+
+BuilderExpr BuilderExpr::max() const {
+  if (!expr_->type()->isNumber() && !expr_->type()->isDateTime()) {
+    throw InvalidQueryError() << "Unsupported type for max aggregate: "
+                              << expr_->type()->toString();
+  }
+  auto agg = makeExpr<AggExpr>(expr_->type(), AggType::kMax, expr_, false, nullptr);
+  auto name = name_.empty() ? "max" : name_ + "_max";
+  return {builder_, agg, name, true};
+}
+
+BuilderExpr BuilderExpr::sum() const {
+  if (!expr_->type()->isNumber()) {
+    throw InvalidQueryError() << "Unsupported type for sum aggregate: "
+                              << expr_->type()->toString();
+  }
+  auto res_type = expr_->type();
+  if (res_type->isInteger() && res_type->size() < 8) {
+    res_type = builder_->ctx_.int64(res_type->nullable());
+  }
+  auto agg = makeExpr<AggExpr>(res_type, AggType::kSum, expr_, false, nullptr);
+  auto name = name_.empty() ? "sum" : name_ + "_sum";
+  return {builder_, agg, name, true};
+}
+
+BuilderExpr BuilderExpr::count(bool is_distinct) const {
+  if (!expr_->is<hdk::ir::ColumnRef>()) {
```

Comment:
Would a scan over the whole table be represented by a column ref? I'm thinking of something similar to `select (*)`.

Reply:
No, column ref is just a reference to a column in a scalar expression. E.g. `SELECT x FROM table;` would have a single column reference in a projection. `SELECT * FROM table;` would require as many column references, as the table has columns.